src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs

```diff
@@ -170,6 +173,35 @@ internal Cci.IFieldReference GetModuleVersionId(Cci.ITypeReference mvidType)
             return _mvidField;
         }
 
+        internal Cci.IFieldReference GetInstrumentationPayload(int analysisKind, Cci.ITypeReference payloadType)
+        {
+            InstrumentationPayloadField payloadField;
+            if (!_instrumentationPayloadFields.TryGetValue(analysisKind, out payloadField))
+            {
+                payloadField = new InstrumentationPayloadField(this, analysisKind, payloadType);
+                if (!_instrumentationPayloadFields.TryAdd(analysisKind, payloadField))
+                {
+                    // Another thread added the field.
+                    payloadField = _instrumentationPayloadFields[analysisKind];
+                }
+            }
+
+            return payloadField;
+        }
+
+        internal ImmutableArray<InstrumentationPayloadField> GetInstrumentationPayloads()
+        {
+            Debug.Assert(IsFrozen);
+            int payloadsLength = _instrumentationPayloadFields.Keys.Max() + 1;
```

Comment:
For reasons that I do not pretend to understand, making the return type IReadOnlyCollection<KVP<>> required an explicit cast in order for the code to compile.